File: php/muse/data_structures/ArrayStack.php
<?php

declare(strict_types=1);

namespace muse\data_structures;

require_once "muse/data_structures/Stack.php";

use muse\data_structures\Stack;

class ArrayStack implements Stack
{
    private const DEFAULT_CAPACITY = 64;

    private $data = null;
    private $logicalSize = 0;
    private $physicalSize = self::DEFAULT_CAPACITY;

    public function __construct(int $physicalSize = 0)
    {
        if ($physicalSize > 1) {
            $this->physicalSize = $physicalSize;
        }
        $this->data = array_pad(array(), $this->physicalSize, 0);
    }

    public function size(): int
    {
        return $this->logicalSize;
    }

    public function isEmpty(): bool
    {
        return $this->logicalSize == 0;
    }

    public function peek(): int
    {
        if ($this->logicalSize == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        return $this->data[$this->logicalSize - 1];
    }

    public function push(int $element): void
    {
        if ($this->logicalSize == $this->physicalSize) {
            $newCapacity = self::DEFAULT_CAPACITY;

            if ($this->physicalSize >= self::DEFAULT_CAPACITY) {
                $newCapacity = $this->physicalSize + ($this->physicalSize >> 1);
            }

            $temp = array_pad(array(), $newCapacity, 0);

            for ($i = 0, $length = $this->logicalSize; $i < $length; $i++) {
                $temp[$i] = $this->data[$i];
            }

            $this->data = $temp;
            $this->physicalSize = $newCapacity;
        }

        $this->data[$this->logicalSize] = $element;

        $this->logicalSize++;
    }

    public function pop(): void
    {
        if ($this->logicalSize == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        $this->logicalSize--;

        $this->data[$this->logicalSize] = null;
    }

    public function capacity(): int
    {
        return $this->physicalSize;
    }

    public function shrink(): void
    {
        $temp = array_pad(array(), $this->logicalSize, 0);

        for ($i = 0, $length = $this->logicalSize; $i < $length; $i++) {
            $temp[$i] = $this->data[$i];
        }

        $this->data = $temp;
        $this->physicalSize = $this->logicalSize;
    }
}

File: python/muse/data_structures/array_stack.py
class ArrayStack[T]:
    DEFAULT_CAPACITY = 64

    def __init__(self, physicalSize: int = 0):
        self.data: list[T | None] = []
        self.logical_size = 0
        self.physical_size = self.DEFAULT_CAPACITY

        if physicalSize > 1:
            self.physical_size = physicalSize
        self.data = [None] * self.physical_size

    def size(self) -> int:
        return self.logical_size

    def is_empty(self) -> bool:
        return self.logical_size == 0

    def peek(self) -> T:
        if self.logical_size == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.data[self.logical_size - 1]

    def push(self, element: T) -> None:
        if self.logical_size == self.physical_size:
            newCapacity = self.DEFAULT_CAPACITY

            if self.physical_size >= self.DEFAULT_CAPACITY:
                newCapacity = self.physical_size + (self.physical_size >> 1)

            temp: list[T | None] = [None] * newCapacity

            for i in range(self.logical_size):
                temp[i] = self.data[i]

            self.data = temp
            self.physical_size = newCapacity

        self.data[self.logical_size] = element

        self.logical_size += 1

    def pop(self) -> None:
        if self.logical_size == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        self.logical_size -= 1

        self.data[self.logical_size] = None

    def capacity(self) -> int:
        return self.physical_size

    def shrink(self) -> None:
        temp: list[T | None] = [None] * self.logical_size

        for i in range(self.logical_size):
            temp[i] = self.data[i]

        self.data = temp
        self.physical_size = self.logical_size
